optimum/onnxruntime/base.py

```diff
@@ -401,64 +419,103 @@ def __init__(
     ):
         super().__init__(session, parent_model)
 
-        if self.use_past:
+        if self.parent_model.use_merged is False and self.use_past is True:
             self.num_pkv = 2
         else:
+            # When using a merged model, we always have the same number of output whether we use past key values or not,
+            # and in the case past key values are used, empty tensors are given as cross-attention past key values as they
+            # are constants
             self.num_pkv = 4
 
-        self.legacy_outputs = any(
+        self.past_key_values_cross_attention_output_names = set()
+        for output_name in self.output_names:
+            if output_name.startswith("present") and "encoder" in output_name:
+                self.past_key_values_cross_attention_output_names.add(output_name)
+
+        self.use_legacy_outputs = self.parent_model.use_merged is False and any(
             "encoder" in output_name and "present" in output_name for output_name in self.output_names
         )
```

Comment:
Instead of rechecking here with the `any(....`), you could just do:

```python
self.use_legacy_outputs = not self.parent_model.use_merged and self.past_key_values_cross_attention_output_names
```